### PR TITLE
Modularize Radial Quadrature Generation

### LINKED ISSUE #12: Modularize radial rules
At present, the implementations of the radial rules hardcode the primitive quadrature rule, and the radial size adjustment. This leads to code duplication, and limits the reusability of the code.

The radial rules should be constructed from an array of nodes and weights. The size adjustment, which consists of scaling the radial nodes by `R` and the weights by `R^3` should likewise happen in a dedicated routine.

### PR DESCRIPTION
**Pull request by wavefunction91**
Unify implementation of radial quadrature generation. Closes #12.

### CODE REVIEW

**Comment by wavefunction91** [comment]
@susilehtola This should add everything that we need to unify the implementations. Dox incoming.

I've written the `traits` for TA, but I can't consolidate the Chebyshev roots with the closed form expression given in https://doi.org/10.1002/jcc.10211 (which also agrees with the Psi4 implementation as far as I can tell). For order `n` Chebyshev quadrature, the points are sinusoidally generated with spacing `pi/n`, where as these implementations generate on `pi/(n+1)`. Unclear how to resolve this. Is there a reason that the TA radial qudature of order `n` uses the Chebyshev nodes of order `n+1`? We just need to be able form the logic to relate the dimension of the base quadrature with that of the transformed quadrature.

File: include/integratorxx/quadratures/gausscheby2.hpp [hunk 1]
```diff
@@ -72,6 +72,6 @@ struct quadrature_traits<GaussChebyshev2<PointType, WeightType>> {
 
     return std::make_tuple(points, weights);
   }
-}
+};
```
Comment (susilehtola) [inline include/integratorxx/quadratures/gausscheby2.hpp:75]: Is this semicolon necessary? Its use was inconsistent between gaysscheby1 and gausscheby2.

**Comment by susilehtola** [comment]
> I've written the `traits` for TA, but I can't consolidate the Chebyshev roots with the closed form expression given in https://doi.org/10.1002/jcc.10211 (which also agrees with the Psi4 implementation as far as I can tell). For order `n` Chebyshev quadrature, the points are sinusoidally generated with spacing `pi/n`, where as these implementations generate on `pi/(n+1)`. Unclear how to resolve this. Is there a reason that the TA radial qudature of order `n` uses the Chebyshev nodes of order `n+1`? We just need to be able form the logic to relate the dimension of the base quadrature with that of the transformed quadrature.

I can't say I fully understand what you are saying, but yes, it looks like there may be an issue with the Psi4 code. I've opened up a ticket https://github.com/psi4/psi4/issues/3004

File: include/integratorxx/quadratures/gausscheby2.hpp (hunk 1 above)
Comment (wavefunction91) [reply include/integratorxx/quadratures/gausscheby2.hpp:75]: It closes `struct quadrature_traits<GaussChebyshev2<...>>`, so it fails to compile otherwise

**Comment by wavefunction91** [comment]
Thanks for looking in to it - glad to know it's not just something I'm missing in my understanding. 

> Gill and Chien's equations are wrong for Mura-Knowles quadrature (they write it with trapezoidal quadrature instead of Chebychev quadrature)

That's what we do too... Interesting, @hjjvandam what does NWChem do? I was able to get (within epsilon) identical grids to NWChem at one point. If there's just an honest discrepancy in the lit / community software, we can expose both via the new modular API.

**Comment by loriab** [comment]
* should https://github.com/wavefunction91/IntegratorXX/blob/master/include/integratorxx/quadratures/gausscheby2.hpp#L26 be scaling by 1/sqrt( 1-x_i^2) ?
* https://github.com/wavefunction91/IntegratorXX/blob/master/include/integratorxx/quadratures/gausscheby1.hpp#L71 denominator quite right: formula is pi/2n and code is 1/(2n pi)

I looked at Susi's equations from Abramowicz a little closer, and the below is what I would do for psi4. The Type1 call is new and the Type2 call isn't actually changing code. I think the scaling for unit form actually _is_ present, but it gets folded into the weight. Of course that doesn't explain why IntegratorXX and Psi4 are producing different values. Could it be something about Integrator looping from `i=0` and filling arrays `weights[i]` with `i+1` and Psi looping from `i=1` and filling arrays `weights[i-1]` with `i`?

Basically, I don't have an answer, so putting forward status from another set of eyes.
```
void RadialGridMgr::getChebychevRootsType1(int n, double r[], double w[]) {
    double piOver2N = M_PI / (2 * n);
    for (int i = 1; i <= n; i++) {
        double x = cos((2 * i - 1.0) * piOver2N);
        r[i - 1] = x;
        w[i - 1] =
            2 * piOver2N * sqrt(1.0 - x * x);  
            // raw weight = pi / n = 2 * piOver2N
            // but scaling for unit form by *= sqrt(1 - x^2), final weight = 2 * piOver2N * sqrt(1 - x^2) = 2 * piOver2N * sin((2 * i - 1.0) * piOver2N)
    }
}

void RadialGridMgr::getChebychevRootsType2(int n, double r[], double w[]) {
    double piOverNPlusOne = M_PI / (n + 1);
    for (int i = 1; i <= n; i++) {
        double x = cos(i * piOverNPlusOne);
        r[i - 1] = x;
        w[i - 1] =
            piOverNPlusOne * sqrt(1.0 - x * x);  
            // raw weight = piOverNPlusOne * sin^2 (i * piOverNPlusOne) = piOverNPlusOne * (1 - cos^2 (i * piOverNPlusOne)) = piOverNPlusOne * (1 - x^2)
            // but scaling for unit form by /= sqrt(1 - x^2), final weight = piOverNplusOne * sqrt(1 - x^2) = piOverNPlusOne * sin(i * piOverNPlusOne)
    }                        
```

**Comment by wavefunction91** [comment]
> - should https://github.com/wavefunction91/IntegratorXX/blob/master/include/integratorxx/quadratures/gausscheby2.hpp#L26 be scaling by 1/sqrt( 1-x_i^2) ?
> - https://github.com/wavefunction91/IntegratorXX/blob/master/include/integratorxx/quadratures/gausscheby1.hpp#L71 denominator quite right: formula is pi/2n and code is 1/(2n pi)

There were bugs in the GC quadratures, fixed in #32 (+ added unit tests). These were place holders before we started the refactor, now they've been validated.

> so putting forward status from another set of eyes.

Those seem to agree with what's currently getting PR-ed (modulo trig optimizations). I *think* this makes things consistent between the two codes for the base quadratures, would be easy enough to check. 

The remaining question is how to validated that the generated Treutler-Ahlrichs (TA) quadratures are consistent. As @susilehtola pointed out, the Gill paper _**might**_ have the wrong final expressions, so I may have been chasing my tail a bit trying to get things to agree. Do we have a third party source to validate the final TA quadratures?

**Comment by loriab** [comment]
Ok, so Psi4 is believed correct at the moment? Whew, but let me know if that changes. I don't have any particular sources on the TA quadratures.

**Comment by wavefunction91** [comment]
If the quadrature generated on [this line](https://github.com/psi4/psi4/blob/afd50aa42403e7980b5e2f1fdf5d8be0f239275a/psi4/src/psi4/libfock/cubature.cc#L2096) is meant to be Chebyshev of the second kind, then yes, this is correct.

**Comment by loriab** [comment]
> If the quadrature generated on [this line](https://github.com/psi4/psi4/blob/afd50aa42403e7980b5e2f1fdf5d8be0f239275a/psi4/src/psi4/libfock/cubature.cc#L2096) is meant to be Chebyshev of the second kind, then yes, this is correct.

Yes, that line is second kind -- I'll fix up the naming and documentation.

**Comment by susilehtola** [comment]
> The remaining question is how to validated that the generated Treutler-Ahlrichs (TA) quadratures are consistent. As @susilehtola pointed out, the Gill paper _**might**_ have the wrong final expressions, so I may have been chasing my tail a bit trying to get things to agree. Do we have a third party source to validate the final TA quadratures?

I can get in touch with the Turbomole developers, since the original implementation of Treutler and Ahlrichs is in that code.

**Comment by wavefunction91** [comment]
@susilehtola I was able to confirm that generating the quadratures via GC-2 -> TA-M4 gave the same thing as the hand coded variant that I got from the Gill paper. Might be good to get some external validation, but it's likely not a priority.

File: include/integratorxx/quadratures/treutlerahlrichs.hpp [hunk 2]
```diff
@@ -115,12 +117,45 @@ struct quadrature_traits<
...
 
-    return std::make_tuple( points, weights );
+};
+
+
+#else
+
+struct TreutlerAhlrichsM4RadialTraits {
 
+  inline static constexpr double alpha = 0.6;
```
Comment (susilehtola) [inline include/integratorxx/quadratures/treutlerahlrichs.hpp:131]: It doesn't seem like we have M3 quadrature, which is the simpler scheme with `alpha=0`

Comment (wavefunction91) [reply include/integratorxx/quadratures/treutlerahlrichs.hpp:131]: Yes, now that things are modular I was planning on adding both in a later PR. We can add them here though, I have no strong opinion.

Comment (susilehtola) [reply include/integratorxx/quadratures/treutlerahlrichs.hpp:131]: Should `alpha=0` be a special case in the code, though?

Comment (wavefunction91) [reply include/integratorxx/quadratures/treutlerahlrichs.hpp:131]: It will be branching logic either way (the Jacobian cares) - whether it's a separate implementation depends on how much we want to expose to the user by default. Will the typical user want to be able to manually specify `alpha` or just specify `M3`/`M4`? My guess is the latter.

**Comment by wavefunction91** [comment]
I've released that we'll need to expose *some* runtime configurability in this modular design. The grids that are exposed right now are relatively simple (the only real configurable parameter is a linear scaling factor), but the DE grids require a bit more configurability (deciding where to cut off the infinite sum, etc). The specification of `alpha` for TA could fall under this type of configurability. 

This kind of configurability will also be useful for folks trying to optimize grids programmatically.

File: include/integratorxx/quadratures/muraknowles.hpp [hunk 3]
```diff
@@ -132,4 +134,35 @@ struct quadrature_traits<
...
+  inline auto radial_jacobian(PointType x) const noexcept {
+    const auto x2 = x*x;
+    return R_ * 3.0 * x2 / (1.0 - x2 * x);
+  }
+
+}; 
+
+template <typename PointType, typename WeightType>
+using MuraKnowles = RadialTransformQuadrature<
+  UniformTrapezoid<PointType,WeightType>,
```
Comment (susilehtola) [inline include/integratorxx/quadratures/muraknowles.hpp:162]: Per our [2022 paper](https://doi.org/10.1063/5.0121187), using an uniform trapezoidal grid gives horrible accuracy for the Mura-Knowles scheme.

I looked through the [Mura-Knowles paper](https://doi.org/10.1063/1.471749) but it was not obvious which rule is supposed to be used. I hope its author @pjknowles can clarify.

Comment (susilehtola) [reply include/integratorxx/quadratures/muraknowles.hpp:162]: According to Peter's email autoreply, he is away from the office until July 13th, so we shouldn't expect a reply until then.

If you already had the Mura-Knowles grid and assumed uniform weights, we can merge this in and open up a separate ticket for the quadrature.

Comment (wavefunction91) [reply include/integratorxx/quadratures/muraknowles.hpp:162]: OK, since this is the standard we've used in our performance benchmarks for GauXC, I'm hesitant to remove it completely. Its clear from your results that we should at least expose the Gauss-Chebyshev variant though. I would just open an issue to track for the time being.

**Comment by hjjvandam** [comment]
Hi David,

NWChem does the following:


      fmn = rm/(1d0+nr)



      do 10 i = 1,nr



       qi = dble(i)/(nr+1d0)

       ri = -alpha*log(1.D0 - qi**rm)

       wi = fmn*alpha*(ri*ri)/(1.D0 - qi**rm)*qi**(rm-1.d0)



       r(i) = ri

       w(i) = wi



   10 continue

where rm=3.0 and alpha=3.3*range where range is the maximum radius covered by the radial grid, and nr is the number of radial grid points. This code lives in nwchem/src/nwdft/xc/xc_murakn.F.

Best wishes,

    Huub


-----------------------------------------------------------------------------------------------------
Hubertus van Dam, 631-344-6020, ***@***.******@***.***>
Brookhaven National Laboratory


From: David Williams-Young ***@***.***>
Date: Friday, July 7, 2023 at 1:59 PM
To: wavefunction91/IntegratorXX ***@***.***>
Cc: Van Dam, Hubertus ***@***.***>, Mention ***@***.***>
Subject: Re: [wavefunction91/IntegratorXX] Modularize Radial Quadrature Generation (PR #27)

Thanks for looking in to it - glad to know it's not just something I'm missing in my understanding.

Gill and Chien's equations are wrong for Mura-Knowles quadrature (they write it with trapezoidal quadrature instead of Chebychev quadrature)

That's what we do too... Interesting, @hjjvandam<https://github.com/hjjvandam> what does NWChem do? I was able to get (within epsilon) identical grids to NWChem at one point. If there's just an honest discrepancy in the lit / community software, we can expose both via the new modular API.

**Comment by wavefunction91** [comment]
Thanks @hjjvandam, that's definitely the uniform trapezoid variant.

File: include/integratorxx/quadratures/gausschebyshev1.hpp [hunk 4]
```diff
@@ -39,8 +39,7 @@ class GaussChebyshev1
   using point_container = typename base_type::point_container;
   using weight_container = typename base_type::weight_container;
 
-  GaussChebyshev1(size_t npts, point_type lo, point_type up)
-      : base_type(npts, lo, up) {}
+  GaussChebyshev1(size_t npts) : base_type(npts) {}
```
Comment (susilehtola) [inline include/integratorxx/quadratures/gausschebyshev1.hpp:42]: This change should really be in a separate commit, no?

File: include/integratorxx/quadratures/gausschebyshev2.hpp [hunk 5]
```diff
@@ -42,8 +42,7 @@ class GaussChebyshev2
   using point_container = typename base_type::point_container;
   using weight_container = typename base_type::weight_container;
 
-  GaussChebyshev2(size_t npts, point_type lo, point_type up)
-      : base_type(npts, lo, up) {}
+  GaussChebyshev2(size_t npts) : base_type(npts) {}
```
Comment (susilehtola) [inline include/integratorxx/quadratures/gausschebyshev2.hpp:45]: same here

File: include/integratorxx/quadratures/gausschebyshev2modified.hpp [hunk 6]
```diff
@@ -43,8 +43,7 @@ class GaussChebyshev2Modified
   using point_container = typename base_type::point_container;
   using weight_container = typename base_type::weight_container;
 
-  GaussChebyshev2Modified(size_t npts, point_type lo, point_type up)
-      : base_type(npts, lo, up) {}
+  GaussChebyshev2Modified(size_t npts) : base_type(npts) {}
```
Comment (susilehtola) [inline include/integratorxx/quadratures/gausschebyshev2modified.hpp:46]: and here

File: include/integratorxx/quadratures/gausschebyshev3.hpp [hunk 7]
```diff
@@ -39,8 +39,7 @@ class GaussChebyshev3
   using point_container = typename base_type::point_container;
   using weight_container = typename base_type::weight_container;
 
-  GaussChebyshev3(size_t npts, point_type lo, point_type up)
-      : base_type(npts, lo, up) {}
+  GaussChebyshev3(size_t npts) : base_type(npts) {}
```
Comment (susilehtola) [inline include/integratorxx/quadratures/gausschebyshev3.hpp:42]: and here

File: include/integratorxx/quadratures/muraknowles.hpp (hunk 3 above)
Comment (pjknowles) [reply include/integratorxx/quadratures/muraknowles.hpp:162]: Here is the code we are using in Molpro for the radial grid evaluation on Mura scheme:
```
     alpha=rsc*bragg(naatom)
     ! assumes that r = - alpha * Log[1- x^idftmr]
     ! suggest use idftmr = 3
     write (6,*) 'alpha, npt, idftmr ',alpha,npt,idftmr
     do ir=1,npt
        qval = dble(ir)/dble(n)
        pt(ir) = -alpha*log(1d0-qval**idftmr)
        wt(ir) =  pt(ir)**2 *alpha*dble(idftmr)* &
             & qval**(idftmr-1)/((1d0-qval**idftmr)*dble(n))
     enddo
     write (6,*) 'pt ',pt(:npt)
     write (6,*) 'wt ',wt(:npt)
```
and `n` is `npt+1`, which I think amounts simple trapezoidal quadrature in the `x` of equation (7) in http://dx.doi.org/10.1063/1.471749, and for the Be atom, this is what is generated:
```
 alpha, npt, idftmr    7.0000000000000000                        48                    3
 pt    5.9499271133939913E-005   4.7600833032953915E-004   1.6066578611874361E-003   3.8089732866334377E-003   7.4413311256047504E-003   1.2863600152286464E-002   2.0437970723343253E-002   3.0529977777593031E-002   4.3509725746882791E-002   5.9753327005997357E-002   7.9644569620570760E-002  0.10357683514824530       0.13195529327830460       0.16519940738611444       0.20374579393428979       0.24805148947543415       0.29859769233640943       0.35589406260236178       0.42048368472012887       0.49294882318131050       0.57391763505319449       0.66407204595981006       0.76415705172765447       0.87499178083047879       0.99748274937600856        1.1326398697743505        1.2815959495630309        1.4456306574120197        1.6262002677869758        1.8249749675453268        2.0438861838399740        2.2851873784253796        2.5515332181111825        2.8460842535535971        3.1726476906379575        3.5358703430273275        3.9415088946346315        4.3968179569776069        4.9111235220893743        5.4966995004301378        6.1701626127089160        6.9548035249514397        7.8847270471719106        9.0128012057250508        10.427589422901853        12.295111781246881        14.988083164786680        19.695799083317340
 wt    6.3191408757371128E-013   1.6178925855460011E-010   4.1478214302682156E-009   4.1457509371301411E-008   2.4736257856667049E-007   1.0652626596586621E-006   3.6641284849845252E-006   1.0694442455644551E-005   2.7541538616163128E-005   6.4278015117145217E-005   1.3857027486190222E-004   2.7986264953446042E-004   5.3525184281221503E-004   9.7758247883188328E-004   1.7164520965350302E-003   2.9130257653514745E-003   4.7998471771280651E-003   7.7072235182964691E-003   1.2098296752658244E-002   1.8615653176199027E-002   2.8143351896444287E-002   4.1889696901313278E-002   6.1498123098954999E-002   8.9196494268593482E-002  0.12799934660813353       0.18198381348671958       0.25666916599665018       0.35954374153109814       0.50080416940484207       0.69440463378479145       0.95956585458830923        1.3229773146815327        1.8220646710507786        2.5099284939617785        3.4629679021480624        4.7929337979063913        6.6665154642672082        9.3381941641618038        13.207427859666060        18.922636048645305        27.580468647622130        41.132941288449700        63.288684869795212        101.72526373304323        174.33040888875871        330.69364765060254        753.72105349957747        2659.6767864598341
```
If more is needed to be sure of what the scheme is, please just ask.

Comment (wavefunction91) [reply include/integratorxx/quadratures/muraknowles.hpp:162]: Thanks @pjknowles, this confirmed that what we had was "canonical". I've added this as reference data in https://github.com/wavefunction91/IntegratorXX/pull/27/commits/8caf841f685dbd3630333a28d34d5bad2a659799

File: include/integratorxx/quadratures/gausschebyshev1.hpp (hunk 4 above)
Comment (wavefunction91) [reply include/integratorxx/quadratures/gausschebyshev1.hpp:42]: Do you mean in a separate PR? The changes are all localized to a single [commit](https://github.com/wavefunction91/IntegratorXX/pull/27/commits/44e6c97c3ab24a169b534408643f68cf45a6c0f2) which is documented in the relevant [issue](https://github.com/wavefunction91/IntegratorXX/issues/42#issue-1795532027), so if the change needs to get reverted, it's trackable. If for some reason this PR doesn't get merged (which is not likely), it can always be cherry-picked as a separate PR.

Since there are quite a few commits that depend on it and I'm not a fan of rewriting git history, I'm going to leave it as is. In general I agree with the sentiment of this concern though, will try to be more disciplined on this in the future.

File: include/integratorxx/quadratures/treutlerahlrichs.hpp (hunk 2 above)
Comment (wavefunction91) [reply include/integratorxx/quadratures/treutlerahlrichs.hpp]: This was changed to a runtime parameter in https://github.com/wavefunction91/IntegratorXX/pull/27/commits/ffad94c7b88dba57ebf5f552938cf2c69fabfd6b and documented in https://github.com/wavefunction91/IntegratorXX/pull/27/commits/5842dbb0ef8ab2ad08bd74d756b4633c65e64bb4. I'm going to call this point resolved if its all the same to you

File: include/integratorxx/quadratures/muraknowles.hpp (hunk 3 above)
Comment (wavefunction91) [reply include/integratorxx/quadratures/muraknowles.hpp:162]: Tracking the `GaussChebyshev2` variant in #51, resolving this conversation for now

File: include/integratorxx/quadratures/gausschebyshev1.hpp (hunk 4 above)
Comment (susilehtola) [reply include/integratorxx/quadratures/gausschebyshev1.hpp:42]: Yes, separate PR, I was probably tired :stuck_out_tongue: 

I need this change for the autogenerated Chebyshev quadrature tests...

**Comment by wavefunction91** [comment]
@susilehtola Anything in particular that would be beneficial to roll up into this PR? I think this is more or less g2g.

**Comment by susilehtola** [comment]
> @susilehtola Anything in particular that would be beneficial to roll up into this PR? I think this is more or less g2g.

I think this is g2g, especially since I want the changes to the Chebyshev routines in master... The modularization can continue in other pulls if necessary
